lib-ee/emqx_ee_connector/src/emqx_ee_connector_tdengine.erl

```diff
@@ -0,0 +1,235 @@
+
+-export([roots/0, fields/1]).
+
+%% `emqx_resource' API
+-export([
+    callback_mode/0,
+    is_buffer_supported/0,
+    on_start/2,
+    on_stop/2,
+    on_query/3,
+    on_batch_query/3,
+    on_get_status/2
+]).
+
+-export([connect/1, do_get_status/1, execute/3]).
+
+-import(hoconsc, [mk/2, enum/1, ref/2]).
+
+-define(TD_HOST_OPTIONS, #{
+    default_port => 6041
+}).
+
+%%=====================================================================
+%% Hocon schema
+roots() ->
+    [{config, #{type => hoconsc:ref(?MODULE, config)}}].
+
+fields(config) ->
+    [
+        {server, server()}
+        | add_default_username(emqx_connector_schema_lib:relational_db_fields(), [])
+    ].
+
+add_default_username([{username, OrigUsernameFn} | Tail], Head) ->
+    Head ++ [{username, add_default_fn(OrigUsernameFn, <<"root">>)} | Tail];
+add_default_username([Field | Tail], Head) ->
+    add_default_username(Tail, Head ++ [Field]).
+
+add_default_fn(OrigFn, Default) ->
+    fun
+        (default) -> Default;
+        (Field) -> OrigFn(Field)
+    end.
+
+server() ->
+    Meta = #{desc => ?DESC("server")},
+    emqx_schema:servers_sc(Meta, ?TD_HOST_OPTIONS).
+
+%%========================================================================================
+%% `emqx_resource' API
+%%========================================================================================
+
+callback_mode() -> always_sync.
+
+is_buffer_supported() -> false.
+
+on_start(
+    InstanceId,
+    #{
+        server := Server,
+        username := Username,
+        password := Password,
+        pool_size := PoolSize
+    } = Config
+) ->
+    ?SLOG(info, #{
+        msg => "starting_tdengine_connector",
+        connector => InstanceId,
+        config => emqx_misc:redact(Config)
+    }),
+
+    {Host, Port} = emqx_schema:parse_server(Server, ?TD_HOST_OPTIONS),
+    Options = [
+        {host, to_bin(Host)},
+        {port, Port},
+        {username, Username},
+        {password, Password},
+        {pool_size, PoolSize},
+        {pool, binary_to_atom(InstanceId, utf8)}
+    ],
+
+    Prepares = parse_prepare_sql(Config),
+    State = maps:merge(Prepares, #{poolname => InstanceId, query_opts => query_opts(Config)}),
+    case emqx_plugin_libs_pool:start_pool(InstanceId, ?MODULE, Options) of
+        ok ->
+            {ok, State};
+        Error ->
+            Error
+    end.
+
+on_stop(InstanceId, #{poolname := PoolName} = _State) ->
+    ?SLOG(info, #{
+        msg => "stopping_tdengine_connector",
+        connector => InstanceId
+    }),
+    emqx_plugin_libs_pool:stop_pool(PoolName).
+
+on_query(InstanceId, {query, SQL}, State) ->
+    do_query(InstanceId, SQL, State);
+on_query(InstanceId, Request, State) ->
+    %% because the `emqx-tdengine` client only supports a whole SQL cmd
+    %% so the `on_query` and `on_batch_query` have the same process, that is:
+    %% we need to contact all data into one SQL cmd and then call the insert API
```

Comment:
If I understood correctly:

```suggestion
    %% because the `emqx-tdengine` client only supports a single SQL cmd
    %% so the `on_query` and `on_batch_query` have the same process, that is:
    %% we need to sollect all data into one SQL cmd and then call the insert API
```

Reply:
Yes, you are right, I want to say a single query string without any params.